CIP-0096? | On-chain dApp Certification Metadata

RSoulatIOHK opened the pull request:
## Abstract
Certification is the process of providing various kinds of assurance of DApps through the provision of verifiable, trustworthy metadata linking to relevant evidence of assurance, such as audit reports, test run data, formal verification proofs etc.

Currently, certification does not have a standardised way to be stored and presented to all stakeholders (DApp developer, DApp stores, end-users, auditors, ...) in an immutable, persistent and verifiable way.

This CIP descibes a standardised method for certificates to be published and stored on-chain and for stakeholders to be able to verify the different claims of this certificate.

This proposal describes how certification metadata can be designed for DApp registration. It first describes the requirements for DApp certification, and then sets out the options for its inclusion in CIP-72.

---

([rendered proposal in branch](https://github.com/RSoulatIOHK/CIPs/blob/cip-certification-metadata/CIP-0096/README.md))

RSoulatIOHK committed:
CIP-123: On-chain Certification Metadata Standard

Initialization of CIP-123 for On-chain Certification Metadata Standard.

This CIP proposes a way to record certification and audits on-chain with mechanisms to:
- check the origin of the certification/audit
- anchor the on-chain certificate to off-chain metadata
- link certificates to registered DApps that used CIP-72
- link certificates to deployed version of DApps

RSoulatIOHK committed:
Fixed a the name of an author and removed a comment

rphair committed:
fixed bracket closure (otherwise GitHub diff unreadable)

rphair committed:
clean up preamble, remove unassigned CIP number

rphair commented:
Thanks @RSoulatIOHK & other authors for the really well developed proposal.

We probably won't be ready for the number `123` for another 4 months or so at the rate we are going.  No worries about needing a CIP number for the related metadata PR since we can also edit that when the CIP number is assigned.

It would also help to avoid confusion if you could rename the CIP folder to something that doesn't have a number in it (e.g. CIP-XXXX, CIP-????, CIP-onchain-cert-metatata).

RSoulatIOHK commented:
Thanks @rphair! We thought we could pick any number that was not used yet, so we picked this one for the three level of certification. Apologies! I'll change the folder name.

RSoulatIOHK committed:
Changed the file name

As requested by @rphair, I changed the folder name

RSoulatIOHK committed:
Addition of the working group minutes as a place of discussion

RSoulatIOHK committed:
Used a markdown linter

Thanks @Niols for the suggestion!

rphair commented:
no worries @RSoulatIOHK - leaving the number assignment to the editors:
https://github.com/cardano-foundation/CIPs/tree/master/CIP-0001#1a-authors-open-a-pull-request
... comes mainly from popular collisions of self-assigned numbers (causing some editorial & social network difficulties) and the tendency to favour small-straight and "yahtzee" numbers in general leading to breaches of a mostly chronological order.  There was also some sentimental value for birth years for honorary figures in Cardano's symbolism and roadmap, but these cases are few.

I can also understand the tendency of working groups to give a CIP draft a simple identified (e.g. 123) and hope that over time some equally simple means of identifying these works in progress can be developed. 🤓

RSoulatIOHK committed:
Added Simon as an author

rphair commented:
@matiwinnetou maybe you would review based on your work in https://github.com/cardano-foundation/CIPs/pull/393?

Ryun1 commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+---
+CIP: ?
+Title: On-chain Certification Metadata Standard
```
```suggestion
Title: On-chain dApp Certification Metadata Standard
```
Since this proposal is aimed at dApp certification, perhaps changing to the title to reflect this may be fitting.

Ryun1 replied:
@rphair has a good eye for naming :)

rphair replied:
@Ryun1 you are absolutely right: this text of this proposal is 100% related to dApps at this time.  @RSoulatIOHK I would also propose that we remove "Standard" from the CIP title, which is implied and sometimes unwanted (since by design CIPs always permit alternatives).  Will retitle it here & in PR title to avoid clutter at the top of the manuscript & hope in advance it's OK with the authors. 🙏

Ryun1 commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+Certification issuers should be free to add additional fields to fit some additional needs.
+
+### List of Certification issuers
```
What's the process for being added to this list? can anyone be added? who verifies them before they are added?

MIxAxIM replied:
Can we apply the same process for adding to the list of Cardano's auditors at CIP 52 here as well? If not, then maybe it is a good time for us to start creating a process to verify or add Cardano's auditors, certification issuers, and testing laboratories to current and future CIPs.

rphair replied:
Cross-referencing https://github.com/cardano-foundation/CIPs/pull/499#pullrequestreview-1386448536.

rphair replied:
Also related to https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1192957854 ... will hold off on further related comments because they all appear related to the suggestion (above & below) about turning this certification table into a structured document & process to be added to that document.

rphair replied:
@Ryun1 @MIxAxIM @RSoulatIOHK - I'm resolving this because it looks like the "list" has been replaced by a process, described in the Specification, for Certification Issuers to publish their own credentials on the chain.  If that's incorrect please respond here. 😎

Ryun1 commented on CIP-onchain-cert-metadata/README.md:
````diff
@@ -0,0 +1,489 @@
+}
+```
+
+The metadata should be discoverable by all certification stakeholders, including end-users, DApp developers, and ecosystem components, such as light wallets and DApp stores. Information should be indexable by certification issuer, DApp developer, DApp and DApp version.
````
> The metadata should be discoverable

How is the off-chain metadata meant to be stored and served to these users?

rphair replied:
@RSoulatIOHK @Ryun1 - I still can't find an elaboration of the current sentence here: https://github.com/RSoulatIOHK/CIPs/blob/cip-certification-metadata/CIP-0096/README.md#off-chain-metadata-format
> **Off-chain Metadata Format**
> The Dapp Certification certificate is complemented by off-chain metadata that can be fetched from off-chain metadata servers.

What are these metadata servers?  Do they already exist & are they described elsewhere?
- if so, can this information be linked here?
- If not, can a rough plan for these servers be put here & included in the Implementation Plan?

Ryun1 reviewed (commented):
Hey @RSoulatIOHK (and other authors), 

I am quite a fan of this proposal! I believe it works nicely with the approach of [CIP-72?](https://github.com/cardano-foundation/CIPs/pull/355). Furthermore I believe this addresses [CPS-01?](https://github.com/cardano-foundation/CIPs/pull/371): by allowing for discoverable, and verifiably correct certification metadata, whilst offering a mechanism for trust.

See my review comments. :)

rphair committed:
add dApp to CIP title

Co-authored-by: Ryan Williams <44342099+Ryun1@users.noreply.github.com>

rphair committed:
remove Standard (redundant) from CIP title

MIxAxIM commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+DApp developers will seek certification from one of the issuers, according to their desired level of certification and will refer to the certificate on several platforms (e.g. on their website or in their DApp registration on a DApp store).
+
+Certification is to be provided by certification issuers including: testing services (level 1), auditors (level 2), and verification services (level 3). Certification issuers will issue these certificates referring to a particular version of a DApp that has succesfully gone through a certification process.
```
The levels of certification, as mentioned in the paragraph, are defined based on the type of certification issuer and the level of testing and verification involved.

Level 1 certification is provided by testing laboratories, which typically involve automated and manual testing to ensure the DApp meets certain functional and security requirements.

Level 2 certification is provided by Cardano's auditors, who perform a more detailed review of the DApp's code, architecture, and security, to ensure that it meets higher standards.

Level 3 certification is provided by Cardano certification issuers and verification services, which typically involve a more comprehensive review of the DApp's design, implementation, and security, and may include formal verification methods to mathematically prove the correctness of the DApp's behavior.

In terms of the differences between auditors and verifiers, auditors typically focus on reviewing the code and implementation of the DApp, while verifiers may also take into account the overall design and architecture of the DApp. Additionally, verification may involve formal methods such as model checking, theorem proving or static analysis are commonly used in verification to rigorously verify the correctness of a DApp's behavior. For example, model checking involves systematically exploring all possible states of a DApp to check if it satisfies certain properties, while theorem proving involves using mathematical logic to prove that a DApp behaves correctly under all possible circumstances. These formal methods are typically not used in auditing, which may rely more on manual review and testing.

Ryun1 replied:
@MIxAxIM 

Thank you for this very well written answer. Perhaps its my lack of understanding - and these are common well known standards?

If these are well known specifications, then linking to them would avoid such questions from less knowledgeable people such as myself. Or if these are levels being defined here then IMO it would be best to include such detail as you have in this answer into the proposal.

RSoulatIOHK replied:
Exactly, they refer to those certification levels:
https://iohk.io/en/blog/posts/2021/10/25/new-certification-levels-for-smart-contracts-on-cardano/

They are yet to be defined properly and we've even refined their definition (in an informal way) during the working group meetings. I can add a reference to a document and add that the actual levels will be standardised in a near future.

rphair replied:
@RSoulatIOHK yes I think it would be a great idea to add this link.  If you added it somewhere please let us know because I can't find it.

rphair replied:
resolving: I can see this is elaborated in the revised text:
> Three levels of certications [sic; I'll submit a correction for this] lare thought of at the time of this CIP: level 1 to 3. They were presented at the following [link](https://iohk.io/en/blog/posts/2021/10/25/new-certification-levels-for-smart-contracts-on-cardano/) and will be fully described in a future document.

MIxAxIM commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+---
+CIP: ?
+Title: On-chain dApp Certification Metadata
```
I don't think "dApp" should be in the title. For example, is an NFT art project that uses just a minting contract considered a dApp too? While dApps are the main users of this CIP, on-chain certification metadata can also be used for other entities. 
Additionally, why was the term "Standard" redundant when it was used in CIP 25, 27, 68, and 71?

rphair replied:
I added it because `dApp` appeared 100% applicable from the proposal as written, but yours is a sensible counterexample & grounds for removing it (I will do so now which also respects the fact that the authors didn't draft it that way).  We really do try to make these titles as specific as possible which is becoming more important as the number of CIPs increases.

Likewise in the early days the word "Standard" was more liberally used, before the discussions that came along in the same year the Governance issue became popular and people began paying far more attention to what's implied by calling something a standard.  I wouldn't like to have it here because it suggests there's only one canonical means of representing on-chain certification data.  For others like the NFT proposals (25, 27) the context makes it more clear that these represent choices of what standard to follow.  Conversely if we keep the word "Standard" in this CIP title then we might as well put it in _every_ CIP title.

MIxAxIM replied:
Since our main focus was and is on DApps, the language and tone of this CIP is DApp-based. However, I believe we can retain the structure while changing some words to include any project. If we do that here, we may consider doing it for CIP 72 as well. The terminology we should use is an open discussion.

Regarding removing the word "Standard," I now understand and completely agree with you. Thank you for the explanation.

rphair replied:
thanks @MIxAxIM - feel free to change the CIP title again without interference when the working group has the particular terminology worked out 😎

rphair replied:
@MIxAxIM with the PR and CIP titles consistent over a month of editing I assume this issue has been resolved. 😎

RSoulatIOHK commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+This CIP descibes a standardised method for certificates to be published and stored on-chain and for stakeholders to be able to verify the different claims of this certificate.
+
+This proposal describes how certification metadata can be designed for DApp registration. It first describes the requirements for DApp certification, and then sets out the options for its inclusion in CIP-72.
```
This is an ancient artifact for when we were considering adding this CIP directly into the CIP 72 for registration. We've decided it would be better to be a separate one. I'll change the sentence.

rphair replied:
Was looking for change but don't see it... @RSoulatIOHK I guess you are still planning on updating this?

rphair replied:
OK, resolving since it looks like offending sentence has been corrected to
> This CIP describes how a certification certificate can be registered on-chain and how to anchor it to a more detailed off-chain certification report.

RSoulatIOHK commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+DApp stores and light wallets will be able to pull DApps information from DApps registration or chain exploring and will link a DApp to a corresponding set of certificates.
+
+End-user will interact with a DApp through a wallet and will be able to check the different certificates obtained by the DApp.
```
I'll write something. Thanks for the suggestion

rphair replied:
@Ryun1 this looks resolved by the paragraph which currently follows the quote:

> One use-case would be for a wallet to maintain a list of valid certificates

RSoulatIOHK commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+### Certification issuers
+
+Certification issuers will broadcast on-chain certificates that will represent th level of certification reached and present evidence of the work done.
+Certification issuers will sign the certificate to attest that they have done the work and prevent certificate forgeries.
```
We had the debate, we thought of either: having the certification issuer publish a public key (as in CIP 52 for example) and add a signature field in the on-chain metadata; or having the certification issuer publish a known wallet that they will use to sign transactions with the metadata. 
We have decided recently to use the second option but this sentence is from the time we were thinking of using the furst one and is not best with our current approach.

rphair replied:
resolving: it looks like this was cleared up in a future round of editing.

RSoulatIOHK commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+The metadata should be discoverable by all certification stakeholders, including end-users, DApp developers, and ecosystem components, such as light wallets and DApp stores. Information should be indexable by certification issuer, DApp developer, DApp and DApp version.
+
+It should be possible for there to be multiple versions of metadata published by the same certification issuer for the same version of a DApp, particularly when a (minor) update of the evidence is necessary. In such a case it should be possible to identify the most recent version of the report for that DApp version. It will also be the case that the same DApp may have certification metadata provi ...
```
In CIP-72, you can register a new update and provide all the script hashes and addresses. If this registration info is up to date, it will be easy to cross check that a particular certification refers to the current version from the on-chain script addresses and hashes that are also provided.

One thing that might be missing in the CIP description is that our goal is only to be able to certify the on-chain components of a DApp.

Ryun1 replied:
> it will be easy to cross check that a particular certification refers to the current version from the on-chain script addresses and hashes that are also provided.

IMO this flow should be detailed in this proposal, as to me this is not clear.

rphair replied:
@RSoulatIOHK if a limitation (disclaimer?) has been placed on the goals of the CIP, I've missed it on the CIP text.  I'm not a dApp expert but what @Ryun1 is saying sounds essential to define the CIP's scope.

rphair replied:
Resolving this because this sentence (currently in Abstract)
>This CIP descibes a standardised method for certificates to be published and stored on-chain and for stakeholders to be able to verify the different claims of this certificate.

only mentions the on-chain components but nothing about certifying anything else.  As with any other CIP, if it's not specifically mentioned then the CIP doesn't cover it.  @RSoulatIOHK I do believe a limitation in scope should have been clarified explicitly here but I'm not going to hold off on a positive review because of it.

RSoulatIOHK committed:
Fixed the schemas

Now the json examples are valid against the schemas and the schemas no longer have errors.

RSoulatIOHK committed:
Fixes

Ryun1 commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+### Suggested validations
+
+- `integrity`: The DApp's metadata off-chain shall match the metadata **anchored** on-chain.
+- `trust`: The DApp's certification metadata transaction shall be signed by a trusted certfication issuer. This means a list of wallets of certification issuers needs to be known. It will then be up to the DApp store to decide which certification issuers are really trusted and publish a list of their own trusted certification issuers.
```
@RSoulatIOHK  - I think this could be a good approach, having this detailed in the proposal is necessary.

rphair replied:
@RSoulatIOHK this is still in an example table in the CIP and I agree there should be an external array (as a file in the CIP directory I assume?) & certification process in the test for being added to that array file.  With your request for status of this CIP I think that would be an essential point for moving forward (cc @Ryun1).

rphair replied:
Resolved (as I checked in another comment) by process for certficiation issuers to publish their own certficiation credentials, as described in the updated Specification: https://github.com/RSoulatIOHK/CIPs/blob/cip-certification-metadata/CIP-0096/README.md#certification-issuers
> Certification issuers will broadcast on-chain certificates that will represent the level of certification reached and present evidence of the work done.
> Certification issuers will sign the transaction with the certification certificate as metadata from a known payment address. They will attest that they have done the work and prevent certificate forgeries.

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+
+`script` an array of objects that represents the whole DApp's on-chain script. Each object is comprised of:
+
+- `fullScriptHash`, a string, is the prefix and script hash or script hash+staking key
```
I would say fullScriptHash is an artefact of history, just put scriptHash, which in fact will a full script hash. I suggest to keep contractAddress as optional, as you can derive it anyway, it is just bech32 format.

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+          "type": "object",
+          "properties": {
+            "twitter": {
+              "type": "string",
```
I suggest to add "pattern" here, see latest version of CIP-72

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+    },
+    "certificationLevel": {
+      "type": "integer",
+      "description": "Describes the action this certification certificate is claiming. For the moment, CERTIFY shall be used to claim a certification that meets the Certfication Standard. AUDIT shall be used to publish on-chain an audit that may not meet the Certification standard. Shall match the on-chain certificationLevel"
```
```
 Shall match the on-chain certificationLevel."
```

rphair replied:
resolving: fixed

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
````diff
@@ -0,0 +1,485 @@
+
+```json
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
````
why draft-04, I believe now there is much newer version, 07 or something

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+                    }
+                }
+            },
+            "fullScriptHash": {
```
I would remove fullScriptHash as mentioned above.

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+                    "description": "Describes the action this certification certificate is claiming. For the moment, CERTIFY shall be used to claim a certification that meets the Certification Standard. AUDIT shall be used to publish on-chain an audit that may not meet the Certification standard"
+                },
+                "certificationLevel": {
+                    "type": "integer",
```
JSON schema allows you nice expresivity with enum: [1, 2, 3] support

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+    "$title": "Certification Certificate", 
+    "type": "object", 
+    "properties": {
+        "subject": {
```
Is there any correlation between this CIP-72's subject?

rphair replied:
Resolving since it seems clear from the following `description` field.  @matiwinnetou if you feel this is still an issue please unresolved & explain. 🤓

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+            },
+            "required": [
+                "scriptHash",
+                "contractAddress"
```
I would not make contractAddress mandatory but optional because it is duplicate of scriptHash, just in bech32 form so anyone can calculate this.

rphair replied:
resolving: not marked as mandatory (anymore)

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+  "subject": "d684512ccb313191dd08563fd8d737312f7f104a70d9c72018f6b0621ea738c5b8213c8365b980f2d8c48d5fbb2ec3ce642725a20351dbff9861ce9695ac5db8",
+  "schemaVersion": 1,
+  "certificationLevel": 1,
+  "certificateIssuer": {
```
Maybe we should have separate CIP for certificate issuers, the reason being is that imagine some of this data changes, the consequence would be that all certificates would have to be resubitted.

rphair replied:
@matiwinnetou cross-referencing https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1160800134 (@Ryun1 @MIxAxIM)

rphair replied:
resolving as per https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1224274902 (i.e. this is a duplicate)

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+  "report": {
+    "reportURLs": [
+        "https://example.io/certificate.pdf",
+        "ipfs://bafybeiemxfal3jsgpdr4cjr3oz3evfyavhwq"
```
I would support only web based IPFS urls but I am not an expert here... is this a valid IPFS url?

rphair replied:
yes, the `ipfs` protocol is recognisable & it's OK for links to have only a CID (i.e. no path): https://docs.ipfs.tech/how-to/address-ipfs-on-web/#native-urls

rphair replied:
resolving since no disputes that the IPFS syntax is OK.

matiwinnetou commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,485 @@
+            "smartContractInfo": {
+                "type": "object",
+                "properties": {
+                    "era": {
```
Why is this necessary at all and if it really is, then it should be a JSON schema enum

RSoulatIOHK committed:
certificationLevel is now enum

Following @matiwinnetou suggestion, I changed the certificationLevel to enum

RSoulatIOHK committed:
Updated the "script" object

Removed fullScriptHash and contract address is no longer mandatory as it can be calculated

RSoulatIOHK committed:
Added some patterns

They have been produced by ChatGPT so they might be incorrect.

RSoulatIOHK committed:
Fixed a missing " in a json file

RSoulatIOHK committed:
Fix the json patterns

RSoulatIOHK committed:
Quick fix

RSoulatIOHK committed:
Added the CDDL schemas

Still work to do:
Regex have not been put into the CDDL
Check the length of the different fields in regex and the spec

RSoulatIOHK committed:
Added a small paragraph on certification level and fixes a typo already spotted

RSoulatIOHK committed:
Update CIP-onchain-cert-metadata/README.md

Co-authored-by: Ryan Williams <44342099+Ryun1@users.noreply.github.com>

RSoulatIOHK committed:
Changed the Copyright notice following @Ryun1 suggestion

Co-authored-by: Ryan Williams <44342099+Ryun1@users.noreply.github.com>

RSoulatIOHK committed:
Fixed missing points in the json schemas descriptions

Thanks matiwinnetou

RSoulatIOHK committed:
Merge branch 'cip-certification-metadata' of https://github.com/RSoulatIOHK/CIPs into cip-certification-metadata

RSoulatIOHK committed:
Fixed the on-chain CDDl and removed the off-chain CDDL as this is not useful

RSoulatIOHK committed:
Various updates (CDDL, Wallet suggestion, Certification level reference, Path to activate addition)

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+### Certification issuers
+
+Certification issuers will broadcast on-chain certificates that will represent th level of certification reached and present evidence of the work done.
```
looks fixed: resolving.

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+### List of Certification issuers
+
+| Certification issuer | URL                | Contact email        | Certification levels | Cardano address |
```
@Ryun1 this also looks pending https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1192957854 i.e. the break-out of this tabular data into a separate, structured file with well-defined fields.  If address were still in that data structure then yes it would have to be elaborated how this address is used.

RSoulatIOHK committed:
Update README.md

rphair commented:
Consensus at CIP meeting today is to promote this to a CIP candidate & assign a CIP number (cc @KtorZ).

Ryun1 reviewed (commented):
Great job on the newest additions to this proposal.

Apart from small details and questions, I think this proposal needs to clearly differentiate itself from definitions of certification levels and how certificate issues gain their authority. Secondly, to me the precise actions around the certificate issuer's credentials and signature are still unclear.

RSoulatIOHK committed:
Update CIP-onchain-cert-metadata/README.md

Spelling mistakes fixed.

Co-authored-by: Ryan Williams <44342099+Ryun1@users.noreply.github.com>

RSoulatIOHK committed:
Update CIP-onchain-cert-metadata/README.md

Spelling and link

Co-authored-by: Ryan Williams <44342099+Ryun1@users.noreply.github.com>

RSoulatIOHK committed:
Removed a section that was supposed to be commented out

Removed a section that was supposed to be commented out

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,543 @@
+---
+CIP: ?
```
```suggestion
CIP: 96
```
Please change number now; it'll be updated in repo top-level `README.md` shortly.  Also rename your directory to `CIP-0096` and update the link in your OP which points to the new pathname of the README file in your branch. 🤓

RSoulatIOHK committed:
Update CIP-onchain-cert-metadata/README.md

Added CIP number

Co-authored-by: Robert Phair <rphair@cosd.com>

RSoulatIOHK committed:
Changed folder name

RSoulatIOHK committed:
Fixed few things

RSoulatIOHK committed:
Addressed the last remaining comments

RSoulatIOHK committed:
Update CIP-0096/README.md

Co-authored-by: Robert Phair <rphair@cosd.com>

RSoulatIOHK committed:
New description for how the off-chain metadata can be hosted.

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+### Custom fields
+
+Certification issuers should be free to add additional fields to fit some additional needs.
```
Resolving; has been elaborated here: https://github.com/RSoulatIOHK/CIPs/blob/cip-certification-metadata/CIP-0096/README.md#aggregators-custom-fields

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+Discussions:
+    - https://github.com/cardano-foundation/CIPs/pull/499
+    - https://github.com/input-output-hk/Certification-working-group/pull/18
+    - https://docs.google.com/document/d/1vC4cS-gMLIvgduSh5Fipn_lU_0Ou0NY4X_a9lWJz2B8
```
Resolving: maybe was a permission / sharing problem, but the URI is valid & readable with no special permissions at this time.

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+## Motivation: why is this CIP necessary?
+
+It is expected that evidence of various kinds of assurance of DApps is recorded in an immutable and verifiable way on the Cardano blockchain.
+Three level of certications are thought of at the time of this CIP: level 1 to 3. They were presented at the following [link](https://iohk.io/en/blog/posts/2021/10/25/new-certification-levels-for-smart-contracts-on-cardano/)
```
@RSoulatIOHK (cc @Ryun1) are you going to move these into a separate CIP on certification levels?  It would make sense to have them defined in something more rigorous than a blog entry almost 2 years old.

Editorially I think it would be fine to leave the certification levels defined here, since as I understand no other CIP refers to them.  They mention the blog entry as a source and are practically confirmed here.  

But please @RSoulatIOHK & other authors let us know if you are going to be producing that "certification levels" CIP while this one is being reviewed... since, if so, those definitions should be in the new CIP rather than in this one.  RSVP 🧐

rphair replied:
Resolved: addressed by https://github.com/cardano-foundation/CIPs/pull/499/commits/520e8f7d1ac96982959d0dd769b4ba9e8e7eb508 (now in `Path to Active`)

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+Level 3 will be used for formal verification. This level is more specialized and is used to bring a full mathematical assurance that critical aspects of the smart contract.
+
+Those level of certification and their standards are yet to be defined. The Cardano Certification Working Group will be one place where those standards are discussed. 
+In order to bootstrap the use of this CIP, an "AUDIT" type with a certification level of 0 has been added in the possible metadata. This will allow auditors and other "certification" issuers to publish on-chain information about the verification that was done on a particular DApp.
```
Resolving, since previously agreed that the original description of a "bootstrapping" level does appear to be of long term use.

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+
+Certification issuers will broadcast on-chain certificates that will represent the level of certification reached and present evidence of the work done.
+
+Certification issuers will sign the transaction with the certification certificate as metadata from a known wallet. They will attest that they have done the work and prevent certificate forgeries.
```
Resolved: text has been changed to "from a known _payment address_"

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+
+<!-- ### List of Certification issuers -->
+
+|<!--  Certification issuer | URL                | Contact email        | Certification levels | Cardano address |
```
Resolved: fixed by eliminating that table altogether.

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+|<!--  Certification issuer | URL                | Contact email        | Certification levels | Cardano address |
+|----------------------|-----               |---------------       |----------------------|------------   |
+| Example Ltd.         | <http://example.com> | contact@example.com  | 1,2                  | EXAMPLEADDRESS   -->
+## Rationale: how does this CIP achieve its goals?
```
Also requested again in https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1224334415.  @RSoulatIOHK - @Ryun1 makes a good point that, if not ready to make a "certification levels" CIP yet, that a paragraph could be added here that another CIP would be desirable / is forthcoming.  If you plan on doing this later then you could add it to your Path to Active that the Certification Team plans to do this at some future point. cc @KtorZ

rphair replied:
Resolved: addressed by https://github.com/cardano-foundation/CIPs/pull/499/commits/520e8f7d1ac96982959d0dd769b4ba9e8e7eb508 (now in `Path to Active`)

RSoulatIOHK committed:
Added that Certification levels needs to be defined for this to become fully active.

rphair commented:
@RSoulatIOHK since https://github.com/cardano-foundation/CIPs/pull/499/commits/cd2e1668694363ac720776e1a02eb63a507d2537 said `Addressed the last remaining comments` I went back to check since I am looking to give this an endorsing review after all the great progress in the last couple of months (p.s. and then again today). 👏

I've checked each one & resolved the many comments that appeared to have been fixed, but according to my reckoning these issues still remain, as of the time of your last commit:
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1160795272
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1195039738
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1195052130

rphair reviewed (commented):
@RSoulatIOHK since https://github.com/cardano-foundation/CIPs/pull/499/commits/cd2e1668694363ac720776e1a02eb63a507d2537 said `Addressed the last remaining comments` I went back to check since I am looking to give this an endorsing review after all the great progress in the last couple of months (p.s.: and also today). 👏

I've resolved the majority of comments that appear fixed, but according to my reckoning these issues still remain at the time of your last commit:
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1160795272
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1195039738
 - https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1195052130

RSoulatIOHK committed:
Addresses the three comments that were left unanswered

Thank you @rphair for your great review!

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+
+`disclaimer`, a string, that contains the legal disclaimer from the certification issuer.
+
+`script` an array of objects that represents the whole DApp's on-chain script. Each object is comprised of:
```
resolved with https://github.com/cardano-foundation/CIPs/pull/499/commits/ec444bf5242a4db5a2f6949d6d89bbc80fcc83ea

rphair commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,548 @@
+
+These certificates are self-standing and can be presented as-is by any stakeholder.
+
+This proposal does not affect any backward compatibility of existing solution but is based on the work done for CIP72 on DApp registration and Discovery. It is also linked to CIP52 on Cardano audit best practices guidelines.
```
resolved with https://github.com/cardano-foundation/CIPs/pull/499/commits/ec444bf5242a4db5a2f6949d6d89bbc80fcc83ea

rphair reviewed (commented):
thanks @RSoulatIOHK for resolving 2 of the 3 questions with https://github.com/cardano-foundation/CIPs/pull/499/commits/ec444bf5242a4db5a2f6949d6d89bbc80fcc83ea, but please how is this question answered?  https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1160795272

rphair commented:
thanks @RSoulatIOHK, that sounds to me like it would be helpful to add, but @Ryun1 works regularly with these types of wallet interactions & it was his question so I believe he's the best choice to clarify his question & accept your answer. 🤔

Ryun1 commented on CIP-onchain-cert-metadata/README.md:
```diff
@@ -0,0 +1,489 @@
+
+It should be possible for there to be multiple versions of metadata published by the same certification issuer for the same version of a DApp, particularly when a (minor) update of the evidence is necessary. In such a case it should be possible to identify the most recent version of the report for that DApp version. It will also be the case that the same DApp may have certification metadata provi ...
+
+It should be possible for wallets to identify to users the certification status of a DApp when they are signing a transaction that is being submitted to a deployed DApp.
```
[RSoulatIOHK](https://github.com/RSoulatIOHK)'s answer comment answer [here](https://github.com/cardano-foundation/CIPs/pull/499#issuecomment-1584798813).

Thanks that makes sense to me - IMO it would be nice to have this more explicit answer in the CIP 😎

Ryun1 commented:
Thanks @rphair @RSoulatIOHK, I have responded to [#499 (comment)](https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1160795272).

RSoulatIOHK committed:
Changed how a wallet could inform a user that they interact with a certified dapp

rphair reviewed (approved):
https://github.com/cardano-foundation/CIPs/pull/499/commits/84a37f4a625596ed47a07db204bd43dc8de9808d takes care of the last question https://github.com/cardano-foundation/CIPs/pull/499#discussion_r1232192297 so I believe this is fully reviewed now... great work everybody 😎

rphair commented:
Thanks @MIxAxIM for dropping in to discuss this at today's CIP editor's meeting.  We noted that there is a ([currently rough draft](https://htmlpreview.github.io/?https://github.com/input-output-hk/Certification-working-group/blob/ac2defc9c47a8437d0b113880df249296c4821b1/Certification%20standard/main.html)) certification document but that including it, or any on-chain implementations of it, would not be required to merge this with `Proposed` status since the _Path to Active_ already refers to it generally. 

Any more rigorous definitions of certification levels that evolve in the course of implementation can be included in a future PR which defines this CIP as `Active`.  There was a suggestion that the Cardano Foundation (mentioning @matiwinnetou) is using a part of this certification method, but we agreed not to require it in _Path to Active_ which only requires implementation in wallets.

One pending item (cc @RSoulatIOHK) that came up at the meeting is how to handle *versioning*.  Consensus (cc @Ryun1) was to try to define certification schema versions in different files (by version ID {1, 2, ...}) like we have for CIP-0072: https://github.com/cardano-foundation/CIPs/tree/master/CIP-0072

RSoulatIOHK committed:
Added the schemas as separated files with versioning

rphair commented:
@RSoulatIOHK @MIxAxIM thanks & the last update works just as well as in CIP-0072.

One thing we may have missed in CIP-0072 (cc @ehanoc @matiwinnetou) which could be a problem again here:  The schema files are duplicated between the CIP text and the associated `.cddl` and `.json` files:
- At least that means that corrections have to be applied in both places & might get out of sync.
- Also we have an issue in both CIPs of the length of the schemas overwhelming the amount of ordinary text.

Could I propose that these schemas simply be linked rather than included inline ([like in CIP-0060](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0060#specification))?  And could the CIP-0072 authors consider doing the same in a near term PR?

@Ryun1 @KtorZ I can imagine there might be reasons to include the whole schemas inline and would like to review them if so.  Pending this one issue it looks like this CIP has met all challenges so far and we might even consider merging it at the next meeting (apply `Last check`?).

RSoulatIOHK committed:
Link to the schema files and removal of the schemas in the README text.

Removed the Schemas from the readme.
Added links to the schemas files.
Added a JSON example for the onchain.
Fixed the subject value in the examples.

RSoulatIOHK committed:
Create level-0.md

RSoulatIOHK committed:
Update level-0.md

RSoulatIOHK committed:
Changes

RSoulatIOHK committed:
Merge branch 'cardano-foundation:master' into cip-certification-metadata

RSoulatIOHK committed:
feat: removed the old descriptions, and provided a more simple overview

rphair commented:
@RSoulatIOHK is this ready to put back on the board for review after your last changes (https://github.com/cardano-foundation/CIPs/pull/499/commits/b4da3432d8ed2e9580bbf8d118f127c2e9871c46)?  As I remember you were looking for some practical implementations before finalising the certification levels.

rphair commented:
@RSoulatIOHK it's a few weeks since https://github.com/cardano-foundation/CIPs/pull/499#issuecomment-2194906117 went unanswered so I'm marking `Waiting for Author`.  Once applying that tag we also want to record what we are waiting for the author _for_... that means please posting the status, and especially indicating contingencies or external factors are behind the continued wait.

rphair commented:
@RSoulatIOHK a lot of work has gone into this, but if this documentation itself is not progressing then I believe it would be best to close it as "abandoned".  If you can update what progress you've making on this issue then please post: otherwise we will close it as such soon.

rphair commented:
(apologies; I hit "close" instead of "comment" button by mistake)

RSoulatIOHK commented:
Hi Robert. Sorry for not answering earlier. We plan to revisit this very soon. We won a Catalyst grant and we'll make drastic changes to the CIP to give it more on chain capabilities.

rphair commented:
@RSoulatIOHK the editors would now be ready mark this as `Likely Abandoned` again (if seen only in terms of time scale)... but since you would be preparing reports for your Catalyst grant mentioned in https://github.com/cardano-foundation/CIPs/pull/499#issuecomment-2372037840 you should have some material to update here: including when the CIP would be completed as part of your development timeline.

In any case to keep this PR open we will need to have more regular updates from you... so please can you mark in your schedule to contribute these every month or two?  If the Catalyst work itself is not progressing then please also let us know so we can close this PR & reopen it once the work continues as expected.  If we don't hear anything at all we would likely have to close this PR with the `Likely Abandoned` tag.

RSoulatIOHK committed:
update: CIP-68 datum

RSoulatIOHK commented:
Hi @rphair. I apologize for the lack of update. We're going to be spending more time and more frequently on this. I just pushed an update to highlight the direction we're taking. This is not ready to be reviewed yet but I want to show that we are indeed thinking about ways to do this certification/evaluation on-chain mechanism.

RSoulatIOHK committed:
removed mentions of the certification working group

RSoulatIOHK committed:
added colll78 suggestion

RSoulatIOHK commented:
> Please add the following to the CIP.

Great suggestion. Check if that's where you wanted it to be added :)

rphair reviewed (commented):
@RSoulatIOHK after 9 more months (3 years total this CIP has been on the drawing board) the editors again need some kind of evidence that this proposed standard is still viable.

Maybe in the meantime there has even been an implementation... as usually required by a "Catalyst grant" (https://github.com/cardano-foundation/CIPs/pull/499#issuecomment-2372037840) especially with the usually rigorous requirements for KPIs and completion metrics?

Considering the span of work that's gone into this submission: rather than closing this as "Abandoned" I would much rather see the document updated as necessary, then this PR pushed back into active review & merged at least as `Proposed`.  If we can't see grounds for doing that before the next "cleanup" cycle (in a few weeks) it would be better to close this PR until perhaps some time in the future when fruitful work can be documented.

rphair commented:
@RSoulatIOHK we are now without your comment on this for over a year, so the editors have to assume you have moved on from this work.  It seems to me the need or this proposal still remains & the pending work would still be useful: but we can't call it a CIP or any other "standard" without confirmation of consistency with practice & update according to peer review.

If you or anyone else working in this field believes this work could be a foundation for a future proposal, please post to request re-opening this or somehow transferring this material into a new working branch.  cc @matiwinnetou @colll78
